File: verifying-an-alien-dictionary/verifying-an-alien-dictionary.py
class Solution:
    def isAlienSorted(self, words: List[str], order: str) -> bool:
        def strCompare(s1, s2, dic):
            for i in range(min(len(s1), len(s2))):
                if dic[s1[i]] < dic[s2[i]]:
                    return True
                elif dic[s1[i]] > dic[s2[i]]:
                    return False

            return len(s1) <= len(s2)
        
        dic = {}
        for i in range(len(order)):
            dic[order[i]] = i
    
        for i in range((len(words) - 1)):
            if not strCompare(words[i], words[i+1], dic):
                return False
        
        return True
        

File: verifying-an-alien-dictionary/verifying-an-alien-dictionary.java
class Solution {
    public boolean isAlienSorted(String[] words, String order) {
        Map<Character, Integer> map = new HashMap<Character, Integer>();
        
        for (int i = 0; i < order.length(); i++) {
            map.put(order.charAt(i), i);
        }
        
        for (int i = 0; i < words.length - 1; i++) {
            if (!isSorted(map, words[i], words[i+1])) {
                return false;
            }
        }
        
        return true;
    }
    
    public boolean isSorted(Map<Character,Integer> map, String str1, String str2) {
        for (int i = 0; i < Math.min(str1.length(), str2.length()); i++) {
            int c1 = map.get(str1.charAt(i));
            int c2 = map.get(str2.charAt(i));
            
            if (c1 < c2) {
                return true;
            } else if (c1 == c2) {
                continue;
            } else {
                return false;
            }
        }
        
        return str1.length() <= str2.length();
    }
}